There are matrix/tensopr failures that need investigation:
```
___ sympy/matrices/expressions/tests/test_applyfunc.py:test_applyfunc_matrix ___
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/utilities/runtests.py", line 1308, in _timeout
    function()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/matrices/expressions/tests/test_applyfunc.py", line 59, in test_applyfunc_matrix
    assert expr.func(*expr.args) == expr
AssertionError
```

And this from tensor

```
_ sympy/tensor/array/tests/test_immutable_ndim_array.py:test_symbolic_indexing _
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/utilities/runtests.py", line 1308, in _timeout
    function()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/tensor/array/tests/test_immutable_ndim_array.py", line 376, in test_symbolic_indexing
    assert Ms[i, j].diff(x) == ImmutableSparseNDimArray([[0, 3], [0, 0]])[i, j]

```

--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -101,6 +101,46 @@ class Expr(Basic, EvalfMixin):
 
         return expr.class_key(), args, exp, coeff
 
+    def __hash__(self):
+        # hash cannot be cached using cache_it because infinite recurrence
+        # occurs as hash is needed for setting cache dictionary keys
+        h = self._mhash
+        if h is None:
+            h = hash((type(self).__name__,) + self._hashable_content())
+            self._mhash = h
+        return h
+
+    def _hashable_content(self):
+        """Return a tuple of information about self that can be used to
+        compute the hash. If a class defines additional attributes,
+        like ``name`` in Symbol, then this method should be updated
+        accordingly to return such relevant attributes.
+        Defining more than _hashable_content is necessary if __eq__ has
+        been defined by a class. See note about this in Basic.__eq__."""
+        return self._args
+
+    def __eq__(self, other):
+        try:
+            other = sympify(other)
+            if not isinstance(other, Expr):
+                return False
+        except (SympifyError, SyntaxError):
+            return False
+        # check for pure number expr
+        if  not (self.is_Number and other.is_Number) and (
+                type(self) != type(other)):
+            return False
+        a, b = self._hashable_content(), other._hashable_content()
+        if a != b:
+            return False
+        # check number *in* an expression
+        for a, b in zip(a, b):
+            if not isinstance(a, Expr):
+                continue
+            if a.is_Number and type(a) != type(b):
+                return False
+        return True
+
     # ***************
     # * Arithmetics *
     # ***************

--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -806,9 +806,9 @@ def disambiguate(*iter):
     for k in mapping:
         # the first or only symbol doesn't get subscripted but make
         # sure that it's a Symbol, not a Dummy
-        k0 = Symbol("%s" % (k), **mapping[k][0].assumptions0)
-        if k != k0:
-            reps[mapping[k][0]] = k0
+        mapk0 = Symbol("%s" % (k), **mapping[k][0].assumptions0)
+        if mapping[k][0] != mapk0:
+            reps[mapping[k][0]] = mapk0
         # the others get subscripts (and are made into Symbols)
         skip = 0
         for i in range(1, len(mapping[k])):

--- a/sympy/core/tests/test_basic.py
+++ b/sympy/core/tests/test_basic.py
@@ -5,7 +5,7 @@ import collections
 import sys
 
 from sympy.core.basic import (Basic, Atom, preorder_traversal, as_Basic,
-    _atomic)
+    _atomic, _aresame)
 from sympy.core.singleton import S
 from sympy.core.symbol import symbols, Symbol
 from sympy.core.function import Function, Lambda
@@ -20,6 +20,12 @@ b1 = Basic()
 b2 = Basic(b1)
 b3 = Basic(b2)
 b21 = Basic(b2, b1)
+
+
+def test__aresame():
+    assert not _aresame(Basic([]), Basic())
+    assert not _aresame(Basic([]), Basic(()))
+    assert not _aresame(Basic(2), Basic(2.))
 
 
 def test_structure():

--- a/sympy/core/tests/test_complex.py
+++ b/sympy/core/tests/test_complex.py
@@ -94,7 +94,7 @@ def test_evalc():
 
 def test_pythoncomplex():
     x = Symbol("x")
-    assert 4j*x == 4*x*I
+    assert 4j*x != 4*x*I
     assert 4j*x == 4.0*x*I
     assert 4.1j*x != 4*x*I
 

--- a/sympy/core/tests/test_numbers.py
+++ b/sympy/core/tests/test_numbers.py
@@ -1687,10 +1687,13 @@ def test_Float_eq():
     assert Float(.75) == S(3)/4
     assert Float(5/18) == 5/18
     # 4473
-    assert t**2 == t**2.0
     assert Float(2.) != 3
     assert Float((0,1,-3)) == S(1)/8
     assert Float((0,1,-3)) != S(1)/9
+    # 16196
+    assert 2 == Float(2)  # as per Python
+    # but in a computation...
+    assert t**2 != t**2.0
 
 
 def test_int_NumberSymbols():

--- a/sympy/core/tests/test_power.py
+++ b/sympy/core/tests/test_power.py
@@ -275,7 +275,7 @@ def test_issue_6100_12942_4473():
     assert x*y != (x*y)**1.0
     # Pow != Symbol
     assert (x**1.0)**1.0 != x
-    assert (x**1.0)**2.0 == x**2
+    assert (x**1.0)**2.0 != x**2
     b = Basic()
     assert Pow(b, 1.0, evaluate=False) != b
     # if the following gets distributed as a Mul (x**1.0*y**1.0 then

--- a/sympy/core/tests/test_priority.py
+++ b/sympy/core/tests/test_priority.py
@@ -123,8 +123,8 @@ def test_sub():
 def test_pow():
     assert h**l == h**x == 1
     assert l**h == x**h == 2
-    assert x**l == 1/x
-    assert l**x == (-1)**x
+    assert (x**l).args == (1/x).args and (x**l).is_Pow
+    assert (l**x).args == ((-1)**x).args and (l**x).is_Pow
 
 
 def test_div():

--- a/sympy/functions/special/tests/test_gamma_functions.py
+++ b/sympy/functions/special/tests/test_gamma_functions.py
@@ -433,8 +433,8 @@ def test_issue_8524():
     assert gamma(e - S.Half).is_positive is False
 
 def test_issue_14450():
-    assert uppergamma(S(3)/8, x).evalf() == uppergamma(0.375, x)
-    assert lowergamma(x, S(3)/8).evalf() == lowergamma(x, 0.375)
+    assert uppergamma(S(3)/8, x).evalf() == uppergamma(S(3)/8, x)
+    assert lowergamma(x, S(3)/8).evalf() == lowergamma(x, S(3)/8)
     # some values from Wolfram Alpha for comparison
     assert abs(uppergamma(S(3)/8, 2).evalf() - 0.07105675881) < 1e-9
     assert abs(lowergamma(S(3)/8, 2).evalf() - 2.2993794256) < 1e-9

--- a/sympy/integrals/tests/test_integrals.py
+++ b/sympy/integrals/tests/test_integrals.py
@@ -294,7 +294,7 @@ def test_issue_3623():
 def test_issue_3664():
     n = Symbol('n', integer=True, nonzero=True)
     assert integrate(-1./2 * x * sin(n * pi * x/2), [x, -2, 0]) == \
-        2*cos(pi*n)/(pi*n)
+        2.0*cos(pi*n)/(pi*n)
     assert integrate(-Rational(1)/2 * x * sin(n * pi * x/2), [x, -2, 0]) == \
         2*cos(pi*n)/(pi*n)
 

--- a/sympy/physics/quantum/tests/test_density.py
+++ b/sympy/physics/quantum/tests/test_density.py
@@ -163,8 +163,8 @@ def test_entropy():
 
     # test for density object
     ent = entropy(d)
-    assert entropy(d) == 0.5*log(2)
-    assert d.entropy() == 0.5*log(2)
+    assert entropy(d) == log(2)/2
+    assert d.entropy() == log(2)/2
 
     np = import_module('numpy', min_module_version='1.4.0')
     if np:

